### DISCUSSION TITLE
pytest life cycle (fixture scoped session)

### DISCUSSION THREAD

**Discussion by MoElleuch** [pytest-dev/pytest#12295] [General]
When using fixtures scoped to session, I initially expected that all fixtures with a session scope would be executed before any test case (for setup) and after all test cases (for teardown). However, I didn't observe the expected behavior. In my first attempt, when I implemented each test case with its own fixture scoped=session, I found that the fixture was executed only for that specific test case. In my second attempt, I tried implementing all fixtures scoped in the file conftest.py i found this behavor :

**INFO     root:conftest.py:7 setup session 5
INFO     root:conftest.py:15 setup session 4**
INFO     root:test_tca1_4.py:12 setup class 4
INFO     root:test_tca1_4.py:19 function setup 4
INFO     root:test_tca1_4.py:28 I'm a test method from a test class 4
INFO     root:test_tca1_4.py:21 function Teardown 4
INFO     root:test_tca1_4.py:14 teardown class 4
INFO     root:conftest.py:17 teardown session 4
INFO     root:conftest.py:10 teardown session 5
**INFO     root:conftest.py:7 setup session 5
INFO     root:conftest.py:15 setup session 4**
INFO     root:test_tca1_5.py:24 setup module 5
INFO     root:test_tca1_5.py:19 setup class 5
INFO     root:test_tca1_5.py:30 function setup 5
INFO     root:test_tca1_5.py:37 function setup 5.1
INFO     root:test_tca1_5.py:45 I'm a test method from a test class 5
INFO     root:test_tca1_5.py:39 function Teardown 5.1
INFO     root:test_tca1_5.py:32 function Teardown 5
INFO     root:test_tca1_5.py:21 teardown class 5
INFO     root:test_tca1_5.py:26 teardown module 5
INFO     root:conftest.py:17 teardown session 4
INFO     root:conftest.py:10 teardown session 5

im using pytest version tested (7.2.0, 7.2.2, 8.0.0) 

**Could anyone help me understand the behavior of pytest and how I can ensure that pytest executes fixtures scoped at the session level only once, at the start of execution and at the end?**

**Comment by The-Compiler** [comment]
That output indeed seems unexpected. Could you please show the code you used for it?

**Reply by MoElleuch** [reply]
this
#conftest.py :

```
import pytest
import logging
import time

@pytest.fixture(autouse=True,scope="session")
def setup_teardown_session_2():
logging.info("setup session 5")
time.sleep(3)
yield
logging.info("teardown session 5")

@pytest.fixture(autouse=True,scope="session")
def setup_teardown_session_3():
logging.info("setup session 4")
yield
logging.info("teardown session 4")
```

under folder tests ihave thoses testcases :

test1.py  :

```
import pytest
import logging
import time
logging.basicConfig(filename='example.log', level=logging.DEBUG, format='%(asctime)s - %(levelname)s - %(message)s')

class TestMyClassTest_4:
    @pytest.fixture(autouse=True,scope="class")
    def setup_teardown_class(self):
        logging.info("setup class 4")
        yield
        logging.info("teardown class 4")
    @pytest.fixture(autouse=True,scope="function")
    def setup_teardown_func(self):
        logging.info("function setup 4")
        yield
        logging.info("function Teardown 4")
    def test_my_test_method_1(self):
        logging.info("I'm a test method from a test class 4")
```

test2.py :

```
import pytest
import logging
import time

logging.basicConfig(filename='example.log', level=logging.DEBUG, format='%(asctime)s - %(levelname)s - %(message)s')

class TestMyClassTest_5:
    @pytest.fixture(autouse=True,scope="class")
    def setup_teardown_class(self):
        logging.info("setup class 5")
        yield
        logging.info("teardown class 5")
    @pytest.fixture(autouse=True,scope="module")
    def setup_teardown_module(self):
        logging.info("setup module 5")
        yield
        logging.info("teardown module 5")
    @pytest.fixture(autouse=True,scope="function")
    def setup_teardown_func(self):
        logging.info("function setup 5")
        yield
        logging.info("function Teardown 5")
    @pytest.fixture(autouse=True,scope="function")
    def setup_teardown_func2(self):
        logging.info("function setup 5.1")
        yield
        logging.info("function Teardown 5.1")
    def test_my_test_method_1(self):
        logging.info("I'm a test method from a test class 5")
```

**Reply by The-Compiler** [reply]
Please [format your code as code properly](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks) and fix its indentation. The idea here is that someone can actually try to reproduce what you're seeing.

**Comment by MoElleuch** [comment]


**Reply by The-Compiler** [reply]
It's been not even 6 hours since your reply, and since then you have:

- pinged me again
- opened an issue that's a duplicate of the discussion

People are doing this in their free time, and have a life outside of this issue tracker as well. Personally, my motivation to look into this is pretty much zero after this.

**Reply by MoElleuch** [reply]
Sorry for the inconvenience caused by the repeated pings and duplicate issue. I'll make sure to be more mindful in the future. Thank you for your understanding.
